core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitUtil.java

```diff
@@ -0,0 +1,208 @@
+          Thread.yield();
+          // Either we have a full commit group, or we have completed writing and need to commit what is left over
+          if (completedRewrites.size() >= rewritesPerCommit || (!running.get() && completedRewrites.size() > 0)) {
+
+            Set<RewriteFileGroup> batch = Sets.newHashSetWithExpectedSize(rewritesPerCommit);
+            for (int i = 0; i < rewritesPerCommit && !completedRewrites.isEmpty(); i++) {
+              batch.add(completedRewrites.poll());
+            }
+
+            try {
+              commitOrClean(batch);
+              committedRewrites.addAll(batch);
+            } catch (Exception e) {
+              LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+            }
+          }
+        }
+      });
+    }
+
+    /**
+     * Places a file group in the queue to be asynchronously committed either when the queue has enough elements
+     * to do a batch of size {@link #rewritesPerCommit} or the service has been closed.
+     * @param group file group to eventually be committed
+     */
+    public void offer(RewriteFileGroup group) {
+      LOG.info("Offered to commit service: {}", group);
+      Preconditions.checkState(running.get(), "Cannot add rewrites to a service which has already been closed");
+      completedRewrites.add(group);
+    }
+
+    /**
+     * Returns all File groups which have been committed
+     */
+    public List<RewriteFileGroup> results() {
+      Preconditions.checkState(committerService.isShutdown(),
+          "Cannot get results from a service which has not been closed");
+      return committedRewrites;
+    }
+
+    @Override
+    public void close() {
+      Preconditions.checkState(running.compareAndSet(true, false),
+          "Cannot close already closed RewriteService");
+      LOG.info("Closing commit service for {}", table);
+      committerService.shutdown();
+
+      try {
+        // All rewrites have completed and all new files have been created, we are now waiting for the commit
+        // pool to finish doing it's commits to Iceberg State. In the case of partial progress this should
+        // have been occurring simultaneously with rewrites, if not there should be only a single commit operation.
+        // In either case this should take much less than 10 minutes to actually complete.
+        if (!committerService.awaitTermination(10, TimeUnit.MINUTES)) {
+          LOG.warn("Commit operation did not complete within 10 minutes of the files being written. This may mean " +
+              "that changes were not successfully committed to the the Iceberg table.");
+        }
+      } catch (InterruptedException e) {
+        throw new RuntimeException("Cannot complete commit for rewrite, commit service interrupted", e);
+      }
+      Preconditions.checkState(completedRewrites.isEmpty(), "File groups offered after service was closed, " +
+          "they were not successfully committed.");
```

Comment:
Nit: I think it's better to move the string to the next line and avoid breaking it across lines.